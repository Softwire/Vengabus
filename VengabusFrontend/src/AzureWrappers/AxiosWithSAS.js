import axios from 'axios';
import moment from 'moment';
import crypto from 'crypto';


/**
 * Provides methods to interact with the api using an SAS token for authentication.
 */
export default class AxiosWithSAS {
    /**
     * @param {string} connectionString The connection string of the server to connect to.
     */
    constructor(connectionString) {
        this.connectionString = connectionString;
    }

    /**
     * Performs a GET request after being passed a url, with the appropriate SAS header.
     * @param {string} url The url that the get request should be sent to.
     * @param {string} config An axios config that is used in the axios request in addition to the SAS authentication.
     * @returns {Promise} The Promise returned by the axios request.
     */
    get = (url, config) => {
        const token = this.getTokenFromConnectionString();
        const sasConfig = this.addAuthToConfig(config, token);
        return axios.get(url, sasConfig);
    }

    /**
     * Performs a DELETE request after being passed a url, with the appropriate SAS header.
     * @param {string} url The url that the get request should be sent to.
     * @param {string} config An axios config that is used in the axios request in addition to the SAS authentication.
     * @returns {Promise} The Promise returned by the axios request.
     */
    delete = (url, config) => {
        const token = this.getTokenFromConnectionString();
        const sasConfig = this.addAuthToConfig(config, token);
        return axios.delete(url, sasConfig);
    }

    /**
     * Performs a PUSH request after being passed a url, with the appropriate SAS header.
     * @param {string} url The url that the get request should be sent to.
     * @param {string} config An axios config that is used in the axios request in addition to the SAS authentication.
     * @returns {Promise} The Promise returned by the axios request.
     */
    push = (url, config) => {
        const token = this.getTokenFromConnectionString();
        const sasConfig = this.addAuthToConfig(config, token);
        return axios.push(url, sasConfig);
    }

    /**
     * Returns an SAS token valid for 1 hour based on the stored connectionString.
     * @returns {object} The SAS token.
     */
    getTokenFromConnectionString() {
        const connectionArray = this.connectionString.split(';');
        let endPoint, sharedAccessKeyName, sharedAccessKey;

        for (let i = 0; i < connectionArray.length; i++) {
            const currentSection = connectionArray[i];

            if (currentSection.startsWith('Endpoint=')) {
                endPoint = 'https:/' + currentSection.slice(13);
            }
            if (currentSection.startsWith('SharedAccessKeyName=')) {
                sharedAccessKeyName = currentSection.slice(20);
            }
            if (currentSection.startsWith('SharedAccessKey=')) {
                sharedAccessKey = currentSection.slice(16);
            }
        }
        return this.createSasToken(endPoint, sharedAccessKeyName, sharedAccessKey);
    }

    /**
     * Generates an SAS token from a given uri, key name, and key.
     * @param {string} uri The URI of the azure service bus.
     * @param {string} keyName The name of the key used to connect to the server, usually "RootManageSharedAccessKey"
     * @param {string} key The key used to connect to the server.
     * @returns {object} The SAS token, valid for 1 hour.
     */
    createSasToken(uri, keyName, key) {
        // Token expires in one hour
        const expiry = moment().add(1, 'hours').unix();

        const string_to_sign = encodeURIComponent(uri) + '\n' + expiry;
        var hmac = crypto.createHmac('sha256', key);
        hmac.update(string_to_sign);
        const signature = hmac.digest('base64');
        const token = 'SharedAccessSignature sr=' + encodeURIComponent(uri) + '&sig=' + encodeURIComponent(signature) + '&se=' + expiry + '&skn=' + keyName;

        return token;
    }

    /**
     * Adds the given SAS token to the headers of the given axios config.
     * @param {object} config The existing axios config.
     * @param {object} token The SAS token to be put in the headers.
     * @returns {object} The config object.
     */
    addAuthToConfig(config, token) {
        const extendedConfig = config || {};
        extendedConfig.headers = extendedConfig.headers || {};
        extendedConfig.headers["Auth-SAS"] = token;
        return extendedConfig;
    }


}