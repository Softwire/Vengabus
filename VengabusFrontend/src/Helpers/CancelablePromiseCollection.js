export class cancelablePromiseCollection {
    constructor() {
        this.trackedPromises = [];
    }


    addNewPromise = (originalPromise,endpointType) => {
        const prom = this.makeCancellable(originalPromise);
        prom.type = endpointType;
        this.trackedPromises.push(prom);
        return prom;
    }

    cancelPromise = (promise) => {
        const index = this.trackedPromises.indexOf(promise);
        if (index > 0) {
            this.trackedPromises.splice(index, 1);
        }
        promise.cancel();
    }

    cancelAllPromises = (endpointType) => {
        if(!endpointType){
            this.trackedPromises.forEach(promise => { promise.cancel();});
            this.trackedPromises = [];
        } else {
            //remove all of type
            const trackedPromisesClone = [...this.trackedPromises];
            trackedPromisesClone.forEach(promise => {
                if (promise.type === endpointType) {
                    this.cancelPromise(promise);
                }
            });
        }
    }

    makeCancellable = (originalPromise) => {
        let hasCancelled = false;

        const wrappedPromise = new Promise((resolve, reject) => {
            originalPromise.then(
                val => hasCancelled ? reject({ isCanceled: true }) : resolve(val),
                error => hasCancelled ? reject({ isCanceled: true }) : reject(error)
            );
        });

        wrappedPromise.cancel = () => { hasCancelled = true; };
        return wrappedPromise;
    };
}