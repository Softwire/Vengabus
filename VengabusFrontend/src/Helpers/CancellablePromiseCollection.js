export class cancellablePromiseCollection {
    constructor() {
        this.trackedPromises = [];
    }


    addNewPromise = (originalPromise,argLabel) => {
        const prom = this.makeCancellable(originalPromise);
        prom.label = argLabel;
        this.trackedPromises.push(prom);
        return prom;
    }

    cancelPromise = (promise) => {
        const index = this.trackedPromises.indexOf(promise);
        if (index > 0) {
            this.trackedPromises.splice(index, 1);
        }
        promise.cancel();
    }

    cancelAllPromises = (labelToCancel) => {
        if(!labelToCancel){
            this.trackedPromises.forEach(promise => { promise.cancel();});
            this.trackedPromises = [];
        } else {
            //remove all of type
            const trackedPromisesClone = [...this.trackedPromises];
            trackedPromisesClone.forEach(promise => {
                if (promise.label === labelToCancel) {
                    this.cancelPromise(promise);
                }
            });
        }
    }

    makeCancellable = (originalPromise) => {
        let hasCancelled = false;

        const wrappedPromise = new Promise((resolve, reject) => {
            originalPromise.then(
                val => hasCancelled ? reject({ isCanceled: true }) : resolve(val),
                error => hasCancelled ? reject({ isCanceled: true }) : reject(error)
            );
        });

        wrappedPromise.cancel = () => { hasCancelled = true; };
        return wrappedPromise;
    };
}