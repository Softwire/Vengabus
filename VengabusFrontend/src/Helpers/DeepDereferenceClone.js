function deepDereferenceClone(obj) {

    //primitives or functions
    function isNonRecursive(obj) {
        return Object(obj) !== obj || typeof obj === 'function';
    }

    if (isNonRecursive(obj)) {
        return obj;
    }

    let ret = Array.isArray(obj) ? [] : {};

    function dfsCopy(sourceNode, cloneNode) {
        //current sourceNode is an array
        if (Array.isArray(sourceNode)) {
            for (let i = 0; i < sourceNode.length; i++) {
                let currentNode = sourceNode[i];
                if (isNonRecursive(currentNode)) {
                    cloneNode.push(currentNode);
                    continue;
                }
                cloneNode.push(Array.isArray(currentNode) ? [] : {});
                dfsCopy(currentNode, cloneNode[i]);
            }
        }
        //current sourceNode is an object
        else {
            let keys = Object.keys(sourceNode);
            for (let i = 0; i < keys.length; i++) {
                let currentNode = sourceNode[keys[i]];
                if (isNonRecursive(currentNode)) {
                    cloneNode[keys[i]] = currentNode;
                    continue;
                }
                cloneNode[keys[i]] = Array.isArray(currentNode) ? [] : {};
                dfsCopy(currentNode, cloneNode[keys[i]]);
            }
        }
    }
    dfsCopy(obj, ret);
    return ret;
}

export { deepDereferenceClone };

/*
This version deals with circular references as well, but that seems like 'a buggy thing to want'.

function deepDereferenceClone(obj) {

    function isNonRecursive(obj) {
        return Object(obj) !== obj || typeof obj === 'function';
    }

    if (isNonRecursive(obj)) {
        return obj;
    }

    let ret = Array.isArray(obj) ? [] : {};
    let sourceStack = [obj];
    let keyStack = ['root'];

    function dfsCopy(sourceNode, cloneNode) {
        //current sourceNode is an array
        if (Array.isArray(sourceNode)) {
            for (let i = 0; i < sourceNode.length; i++) {
                let currentNode = sourceNode[i];
                if (isNonRecursive(currentNode)) {
                    cloneNode.push(currentNode);
                    continue;
                }
                let cyclicRefIndex = sourceStack.indexOf(currentNode);
                if (cyclicRefIndex >= 0) {
                    cloneNode.push("Cyclic reference detected: " + keyStack.join('.') + '[' + i + ']  is a reference to ' + keyStack.slice(0, cyclicRefIndex + 1).join('.'));
                    continue;
                }
                cloneNode.push(Array.isArray(currentNode) ? [] : {});
                sourceStack.push(currentNode);
                keyStack.push('[' + i + ']');
                dfsCopy(currentNode, cloneNode[i]);
                keyStack.pop();
                sourceStack.pop();
            }
        }
        //current sourceNode is an object
        else {
            let keys = Object.keys(sourceNode);
            for (let i = 0; i < keys.length; i++) {
                let currentNode = sourceNode[keys[i]];
                if (isNonRecursive(currentNode)) {
                    cloneNode[keys[i]] = currentNode;
                    continue;
                }
                let cyclicRefIndex = sourceStack.indexOf(currentNode);
                if (cyclicRefIndex >= 0) {
                    cloneNode[keys[i]] = "Cyclic reference detected: " + keyStack.join('.') + '.' + keys[i] + ' is a reference to ' + keyStack.slice(0, cyclicRefIndex + 1).join('.');
                }
                cloneNode[keys[i]] = Array.isArray(currentNode) ? [] : {};
                sourceStack.push(currentNode);
                keyStack.push(keys[i]);
                dfsCopy(currentNode, cloneNode[keys[i]]);
                keyStack.pop();
                sourceStack.pop();
            }
        }
    }
    dfsCopy(obj, ret);
    return ret;
}
*/