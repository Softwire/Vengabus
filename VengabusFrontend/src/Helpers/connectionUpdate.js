class ConnectionUpdate {
    constructor() {
        this.updateCallBacks = [];
    }

    registerForUpdates(callback) {
        this.updateCallBacks.push(callback);
    }

    deregisterForUpdates(callback) {
        this.updateCallBacks = this.switchCallbacks.filter((cb) => cb !== callback);
    }

    performUpdate() {
        this.updateCallBacks.forEach((registeredCallback) => {
            registeredCallback();
        });
    }
}
export const connectionUpdate = new ConnectionUpdate();
