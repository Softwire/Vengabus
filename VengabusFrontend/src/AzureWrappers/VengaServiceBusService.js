import { AxiosWithSAS } from './AxiosWithSAS';
require('util.promisify').shim();

/*
    This class exists as our own wrapper around the equivalent Azure API object,
    so that we can make any appropriate improvements to its API.
    e.g. Converting everything to promises.  
*/
export class VengaServiceBusService {
    constructor(connectionString, apiRoot) {
        this.axiosWithSAS = new AxiosWithSAS(connectionString);
        this.apiRoot = apiRoot;
        this.jsonConfig = {
            headers: {
                "Content-Type": "application/json"
            }
        };
    }

    /**
     * Gets the details of all queues in the current namespace from the server.
     * @return {object} The queues returned by the server.
     */
    listQueues = () => {
        const url = this.apiRoot + 'queues';
        return this.axiosWithSAS.get(url);
    }

    /**
     * Gets the details of a particular queue.
     * @param {string} queueName The name of the queue to get details for.
     * @return {object} The queues returned by the server.
     */
    getQueueDetails = (queueName) => {
        const url = this.apiRoot + `queues/${queueName}`;
        return this.axiosWithSAS.get(url);
    }

    /**
     * Updates the properties of a queue
     * @param {object} queueDescription Object containing the new properties of the queue. It is specified in this which queue we are updating.
     * For more information on the form of this object see https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.messaging.queuedescription?view=azure-dotnet
     */
    updateQueue = (queueDescription) => {
        const url = this.apiRoot + 'queues/update';
        const config = this.jsonConfig;
        this.axiosWithSAS.post(url, queueDescription, config);
    }

    /**
     * Renames a queue
     * @param {string} oldName The original name of the queue.
     * @param {string} newName The new name of the queue.
     */
    renameQueue = (oldName, newName) => {
        const url = this.apiRoot + 'queues/rename';
        const config = this.jsonConfig;
        const names = {
            oldName: oldName,
            newName: newName
        };
        this.axiosWithSAS.post(url, names, config);
    }

    /**
     * Deletes a queue
     * @param {string} name Name of the queue to be deleted.
     */
    deleteQueue = (name) => {
        const url = this.apiRoot + 'queues/delete/' + name;
        const config = this.jsonConfig;
        this.axiosWithSAS.post(url, null, config);
    }

    /**
     * Gets the details of all topics in the current namespace from the server.
     * @return {object} The queue returned by the server.
     */
    listTopics = () => {
        const url = this.apiRoot + 'topics';
        return this.axiosWithSAS.get(url);
    }

    /**
     * Gets the details of a particular topic.
     * @param {string} topicName The name of the topic to get details for.
     * @return {object} The topic returned by the server.
     */
    getTopicDetails = (topicName) => {
        const url = this.apiRoot + `topics/${topicName}`;
        return this.axiosWithSAS.get(url);
    }

    /**
     * Updates the properties of a queue
     * @param {object} topicDescription Object containing the new properties of the topic. It is specified in this which topic we are updating.
     * For more information on the form of this object see https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.messaging.topicdescription?view=azure-dotnet
     */
    updateTopic = (topicDescription) => {
        const url = this.apiRoot + 'topics/update';
        const config = this.jsonConfig;
        this.axiosWithSAS.post(url, topicDescription, config);
    }

    /**
     * Renames a topic
     * @param {string} oldName The original name of the topic.
     * @param {string} newName The new name of the topic.
     */
    renameTopic = (oldName, newName) => {
        const url = this.apiRoot + 'topics/rename';
        const config = this.jsonConfig;
        const names = {
            oldName: oldName,
            newName: newName
        };
        this.axiosWithSAS.post(url, names, config);
    }

    /**
     * Deletes a topic
     * @param {string} name Name of the topic to be deleted.
     */
    deleteTopic = (name) => {
        const url = this.apiRoot + 'topics/delete/' + name;
        const config = this.jsonConfig;
        this.axiosWithSAS.post(url, null, config);
    }

    /**
     * Gets the details of all subscriptions in a given topic from the server.
     * @param {string} topicName The name of the topic to get subscriptions from.
     * @return {object} The subsctiptions returned by the server.
     */
    listSubscriptions = (topicName) => {
        const url = this.apiRoot + `subscriptions/${topicName}`;
        return this.axiosWithSAS.get(url);
    }


    /**
     * Gets the details of a particular topic.
     * @param {string} parentTopicName The name of the parent topic for the subscription.
     * @param {string} subscriptionName The name of the subscription to get details for.
     * @return {object} The topic returned by the server.
     */
    getSubscriptionDetails = (parentTopicName, subscriptionName) => {
        const url = this.apiRoot + `subscriptions/${parentTopicName}/${subscriptionName}`;
        return this.axiosWithSAS.get(url);
    }

    /**
     * Updates the properties of a subscription
     * @param {object} subDescription Object containing the new properties of the subscription. It is specified in this which subscription we are updating.
     * For more information on the form of this object see https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.messaging.subscriptiondescription?view=azure-dotnet
     */
    updateSubscription = (subDescription) => {
        const url = this.apiRoot + 'subscriptions/update';
        const config = this.jsonConfig;
        this.axiosWithSAS.post(url, subDescription, config);
    }

    /**
     * Deletes a subscription
     * @param {string} name Name of the subscription to be deleted.
     * @param {string} parentTopic Name of the paretn topic of the subscription.
     */
    deleteSubscription = (name, parentTopic) => {
        const url = this.apiRoot + 'subscriptions/delete/' + name;
        const config = this.jsonConfig;
        this.axiosWithSAS.post(url, parentTopic, config);
    }

    /**
     * Sends a message to the given queue on the server.
     * @param {string} queueName The name of the queue to send the message to.
     * @param {object} message The message to send to the queue, in the format:
     * {
     *      "customProperties": {},
     *      "predefinedProperties": {},
     *      "messageBody": "string"
     * }
     */
    sendMessageToQueue = (queueName, message) => {
        const url = this.apiRoot + `queues/${queueName}/messages`;
        const config = this.jsonConfig;
        this.axiosWithSAS.post(url, message, config);
    }

    /**
     * Sends a message to the given topic on the server.
     * @param {string} topicName The name of the topic to send the message to.
     * @param {object} message The message to send to the queue, in the format:
     * {
     *      "customProperties": {},
     *      "predefinedProperties": {},
     *      "messageBody": "string"
     * }
     */
    sendMessageToTopic = (topicName, message) => {
        const url = this.apiRoot + `topics/${topicName}/messages`;
        const config = this.jsonConfig;
        this.axiosWithSAS.post(url, message, config);
    }

    purgeQueueMessages = (queueName) => {
        const url = this.apiRoot + `queues/${queueName}/messages`;
        return this.axiosWithSAS.delete(url);
    }

    deleteQueueSingleMessage = (queueName, messageId, uniqueId) => {
        const messageIdEncoded = encodeURIComponent(messageId);
        const url = this.apiRoot + `queues/${queueName}/messages/${uniqueId}?messageId=${messageIdEncoded}`;
        return this.axiosWithSAS.delete(url);
    }

    deleteQueueSingleDeadLetterMessage = (queueName, messageId, uniqueId) => {
        const messageIdEncoded = encodeURIComponent(messageId);
        const url = this.apiRoot + `queues/${queueName}/deadletters/${uniqueId}?messageId=${messageIdEncoded}`;
        return this.axiosWithSAS.delete(url);
    }

    purgeTopicMessages = (topicName) => {
        const url = this.apiRoot + `topics/${topicName}/messages`;
        return this.axiosWithSAS.delete(url);
    }

    deleteTopicSingleMessage = (topicName, messageId, uniqueId) => {
        const messageIdEncoded = encodeURIComponent(messageId);
        const url = this.apiRoot + `topics/${topicName}/messages/${uniqueId}?messageId=${messageIdEncoded}`;
        return this.axiosWithSAS.delete(url);
    }

    purgeSubscriptionMessages = (topicName, subscriptionName) => {
        const url = this.apiRoot + `subscriptions/${topicName}/${subscriptionName}/messages`;
        return this.axiosWithSAS.delete(url);
    }

    deleteSubscriptionSingleMessage = (topicName, subscriptionName, messageId, uniqueId) => {
        const messageIdEncoded = encodeURIComponent(messageId);
        const url = this.apiRoot + `subscriptions/${topicName}/${subscriptionName}/messages/${uniqueId}?messageId=${messageIdEncoded}`;
        return this.axiosWithSAS.delete(url);
    }

    deleteSubscriptionSingleDeadLetterMessage = (topicName, subscriptionName, messageId, uniqueId) => {
        const messageIdEncoded = encodeURIComponent(messageId);
        const url = this.apiRoot + `subscriptions/${topicName}/${subscriptionName}/deadletters/${uniqueId}?messageId=${messageIdEncoded}`;
        return this.axiosWithSAS.delete(url);
    }

    listQueueMessages = (queueName, messageCount) => {
        let queryString = "";
        if (messageCount) {
            queryString = "/?messageCount=" + messageCount;
        }
        const url = this.apiRoot + `queues/${queueName}/messages` + queryString;
        return this.axiosWithSAS.get(url);
    }

    listSubscriptionMessages = (topicName, subscriptionName, messageCount) => {
        let queryString = "";
        if (messageCount) {
            queryString = "/?messageCount=" + messageCount;
        }
        const url = this.apiRoot + `subscriptions/${topicName}/${subscriptionName}/messages` + queryString;
        return this.axiosWithSAS.get(url);
    }


    listQueueDeadLetterMessages = (queueName, messageCount) => {
        let queryString = "";
        if (messageCount) {
            queryString = "/?messageCount=" + messageCount;
        }
        const url = this.apiRoot + `queues/${queueName}/deadletters` + queryString;
        return this.axiosWithSAS.get(url);
    }

    listSubscriptionDeadLetterMessages = (topicName, subscriptionName, messageCount) => {
        let queryString = "";
        if (messageCount) {
            queryString = "/?messageCount=" + messageCount;
        }
        const url = this.apiRoot + `subscriptions/${topicName}/${subscriptionName}/deadletters` + queryString;
        return this.axiosWithSAS.get(url);
    }

    getQueueMostRecentDeadletter = (queueName) => {
        const url = this.apiRoot + `queues/${queueName}/mostRecentDeadletter`;
        return this.axiosWithSAS.get(url);
    }

    getSubscriptionMostRecentDeadletter = (topicName, subscriptionName) => {
        const url = this.apiRoot + `subscriptions/${topicName}/${subscriptionName}/mostRecentDeadletter`;
        return this.axiosWithSAS.get(url);
    }



    /**
     * Returns the pre-defined properties that can be read from messages.
     * @returns {string[]} The available properties.
     */
    getReadableMessageProperties = () => {
        const url = this.apiRoot + 'messages/properties/readable';
        return this.axiosWithSAS.get(url);
    }

    /**
     * Returns the pre-defined properties that can be set to messages.
     * @returns {string[]} The allowed properties.
     */
    getWriteableMessageProperties = () => {
        const url = this.apiRoot + 'messages/properties/writeable';
        return this.axiosWithSAS.get(url);
    }
}
