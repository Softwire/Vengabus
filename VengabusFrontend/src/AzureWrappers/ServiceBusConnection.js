import { VengaServiceBusService } from './VengaServiceBusService';

class ServiceBusConnection {
    constructor() {
        this.activeServiceBusConString = '';
        this.activeAPIroot = '';
    }

    /**
     * Updates the Azure connection string used to connect to the service bus.
     * @param {string} newConnectionString The new connection string to use.
     * Should be in the format generated by the Azure portal:
     * Endpoint=sb://<EndpointLocation>/;SharedAccessKeyName=<KeyName>;SharedAccessKey=<SharedAccessKey>
     * e.g.
     * Endpoint=sb://vengabusdemo.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SGd6kBFkdDBGKs5FIBSgibSgSG5dsUBGS+8=
     */
    setConnectionString = (newConnectionString) => {
        this.activeServiceBusConString = newConnectionString;
    };

    /**
     * Updates the location of the API backend to be used by the ServiceBusConnection.
     * http:// and a trailing slash will be added if omitted.
     * @param {string} newURI The new API root URI to use.
     */
    setApiRoot = (newURI) => {
        this.activeAPIroot = this.formatApiRootString(newURI);
    };

    /**
     * Takes a string and prepends http:// and adds a trailing slash if not already present.
     * @param {string} APIRootString The string to format.
     * @return {string} The formatted string.
     */
    formatApiRootString = (APIRootString) => {
        if (!APIRootString.startsWith("http")) { //Catches both http and https cases
            APIRootString = "http://" + APIRootString;
        }
        if (!APIRootString.endsWith("/")) {
            APIRootString = APIRootString + "/";
        }
        return APIRootString;
    };

    /**
     * Returns a VengaServiceBusService that uses the current connection string and API root location.
     * @return {VengaServiceBusService} The VengaServiceBusService defined by the settings in this object.
     */
    getServiceBusService = () => {
        try {
            return new VengaServiceBusService(this.activeServiceBusConString, this.activeAPIroot);
        } catch (err) {
            // Obviously we'd prefer this didn't happen, but it can easily occur if, e.g. ConnStrings aren't provided yet.
            // Failing to create a connection shouldn't be an immediate fatal error ... but *USING* a failed connection probably should.
            // So don't error - just return `undefined`.
            console.error(err);
            return undefined;
        }
    };
}
export const serviceBusConnection = new ServiceBusConnection();
