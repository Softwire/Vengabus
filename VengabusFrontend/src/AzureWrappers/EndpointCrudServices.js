import { serviceBusConnection } from './ServiceBusConnection';

/*
  Common Interface:
  {
    rename: function(oldName, newName),
    update: function(endpointData),
    delete: function(name)
  }
 */

export class QueueCrudService {
    constructor() {
        this.serviceBusService = serviceBusConnection.getServiceBusService();

        this.rename = this.serviceBusService.renameQueue;
        this.update = this.serviceBusService.updateQueue;
        this.delete = this.serviceBusService.deleteQueue;
    }
}

export class TopicCrudService {
    constructor() {
        this.serviceBusService = serviceBusConnection.getServiceBusService();

        this.rename = this.serviceBusService.renameTopic;
        this.update = this.serviceBusService.updateTopic;
        this.delete = this.serviceBusService.deleteTopic;
    }
}

export class SubscriptionCrudService {
    constructor(parentTopic) {
        this.serviceBusService = serviceBusConnection.getServiceBusService();
        this.parentTopic = parentTopic;

        this.rename = () => console.error('cannot rename subscriptions because #Microsoft');
        this.update = this.serviceBusService.updateSubscription;
        this.delete = (subscriptionName) => this.serviceBusService.deleteSubscription(subscriptionName, this.parentTopic);
    }
}
