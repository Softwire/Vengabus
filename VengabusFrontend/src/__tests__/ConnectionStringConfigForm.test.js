import 'jest-localstorage-mock';
import { mount, configure } from 'enzyme';
import Adaptor from 'enzyme-adapter-react-16';
import { testHelper } from '../Helpers/testHelper';

import React from 'react';
import {
  ConnectionStringConfigForm,
  LOCAL_STORAGE_STRINGS
} from "../Components/ConnectionStringConfigForm";
import { serviceBusConnection } from '../AzureWrappers/ServiceBusConnection';

configure({ adapter: new Adaptor() });

it('component renders fine when no localStorage is present', () => {
    localStorage.setItem(LOCAL_STORAGE_STRINGS.ConnectionString, undefined);
    const wrapper = mount(<ConnectionStringConfigForm />);
});

it('component renders fine when localStorage *is* present', () => {
    localStorage.setItem(LOCAL_STORAGE_STRINGS.ConnectionString, 'some Value');
    const wrapper = mount(<ConnectionStringConfigForm />);
});

it('localStore is updated when the form is changed', () => {
    localStorage.setItem(LOCAL_STORAGE_STRINGS.ConnectionString, 'before');
    const wrapper = mount(<ConnectionStringConfigForm />);

    const connectionStringInput = wrapper.find('#connectionString');
    connectionStringInput.simulate('change', { target: { value: 'after' } });

    return testHelper.afterReactHasUpdated().then(() => {
        expect(localStorage.getItem(LOCAL_STORAGE_STRINGS.ConnectionString)).toEqual('after');
    });
});

it('ServiceBusConnection is updated appropriately on page load', () => {
    serviceBusConnection.setConnectionString('before');
    localStorage.setItem(LOCAL_STORAGE_STRINGS.ConnectionString, 'after');

    mount(<ConnectionStringConfigForm />);
    return testHelper.afterReactHasUpdated().then(() => {
        expect(serviceBusConnection.activeServiceBusConString).toEqual('after');
    });
});
