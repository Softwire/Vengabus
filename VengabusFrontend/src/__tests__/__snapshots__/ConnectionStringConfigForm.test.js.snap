// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionStringConfigForm renders correctly 1`] = `
.emotion-0 {
  color: black;
  margin: 5px;
}

.emotion-3 {
  padding: 5px;
}

.emotion-2 {
  color: black;
  overflow-wrap: break-word;
}

.emotion-1 {
  background: #66b3ff;
}

<form
  className="emotion-3"
>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="connectionString"
    >
      ServiceBus Connection String
    </label>
    <input
      className="form-control"
      id="connectionString"
      onChange={[Function]}
      placeholder="Enter Connection String"
      type="text"
      value="connStringInLocalStorage"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="APILocationForm"
    >
      ServiceBus API Server String
    </label>
    <input
      className="form-control"
      id="APILocationForm"
      onChange={[Function]}
      placeholder="API Location, e.g. http://a.com:63591/"
      type="text"
      value="UnformattedAPIRootInLocalStorage"
    />
  </div>
  <button
    className="emotion-0 btn btn-default"
    disabled={false}
    id="connectButton"
    onClick={[Function]}
    type="button"
  >
    Connect
  </button>
  <div>
    <br />
  </div>
  <div
    className="panel panel-default"
  >
    <div
      className="emotion-1 panel-heading"
    >
      ServiceBus Details
    </div>
    <div
      className="emotion-2 panel-body"
    >
      <div>
        Your Connection string: connStringInLocalStorage
      </div>
      <div>
        Name:  
      </div>
      <div>
        Location:  
      </div>
      <div>
        Status:  
      </div>
      <div>
        Permissions:  
      </div>
    </div>
  </div>
</form>
`;
